lib/resty/upstream/healthcheck.lua

```diff
@@ -92,6 +93,92 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
     end
 end
 
+local function get_unchecked_peers(peers)
+    local peers_to_check = {}
+    for _, peer in ipairs(peers) do
+        if not checked_peers_upstream[peer.name] then
+            table.insert(peers_to_check, peer)
+        end
+    end
+    return peers_to_check
+end
+
+local function get_checked_peers(peers)
+    local checked_peers = {}
+    for _, peer in ipairs(peers) do
+        if checked_peers_upstream[peer.name] then
+            table.insert(checked_peers, peer)
+        end
+    end
+    return checked_peers
+end
+
+local function get_all_peers(u)
+    local primary_peers, err = upstream.get_primary_peers(u)
+    if not primary_peers then
+        errlog("failed to get primary peers: ", err)
+        return
+    end
+
+    local backup_peers, err = upstream.get_backup_peers(u)
+    if not backup_peers then
+        errlog("failed to get backup peers: ", err)
+        return
+    end
+
+    return primary_peers, backup_peers
+end
+
+local function get_all_unchecked_peers(u)
+    local primary_peers, backup_peers = get_all_peers(u)
+
+    local ppeers_to_check = get_unchecked_peers(primary_peers)
+    local bpeers_to_check = get_unchecked_peers(backup_peers)
+
+    return ppeers_to_check, bpeers_to_check
+end
+
+local function get_all_checked_peers(u)
+    local primary_peers, backup_peers = get_all_peers(u)
+
+    local ppeers_checked = get_checked_peers(primary_peers)
+    local bpeers_checked = get_checked_peers(backup_peers)
+
+    return ppeers_checked, bpeers_checked
+end
+
+local function set_status(peers, is_backup, ctx)
+    local dict = ctx.dict
+    local u = ctx.upstream
+
+    for i = 1, #peers do
+        local peer = peers[i]
+        local already_set = dict:get(gen_peer_key("ok:", u, is_backup, peer.id))
+        if already_set then
+            return
+        end
+
+        local checked_peer_upstream = checked_peers_upstream[peer.name]
+
+        local oks, _ = dict:get(gen_peer_key("ok:", checked_peer_upstream, is_backup, peer.id))
+        if oks then
+            dict:set(gen_peer_key("ok:", u, is_backup, peer.id), 1)
+        end
+
+        local noks, _ = dict:get(gen_peer_key("nok:", checked_peer_upstream, is_backup, peer.id))
+        if noks then
+            dict:set(gen_peer_key("nok:", u, is_backup, peer.id), 1)
+        end
+    end
+end
+
+local function copy_status(ctx)
+    local ppeers_checked, bpeers_checked = get_all_checked_peers(ctx.upstream)
+
+    set_status(ppeers_checked, false, ctx)
+    set_status(bpeers_checked, true, ctx)
+end
```

Comment:
This function is the only place you call `get_all_checked_peers`, which is the only caller of `get_checked_peers`. Again, this could be simplified by inlining the ... uh ... check into `set_status`, which should probably be called `copy_status_if_checked`:
```lua
local function copy_status(ctx)
  local ppeers, bpeers = get_all_peers(ctx.upstream)
  copy_status_if_checked(ppeers, false, ctx)
  copy_status_if_checked(bpeers, true, ctx)
end
```